// @pdftron/webviewer/scripts/optimize-script-helper.mjs
import path from 'path';
import { fileURLToPath } from 'url';

// `__dirname` is only used by `wvResolve` and Salesforce path handling, both of which are exercised by Node script consumers — never by Vitest browser tests. Compute it lazily and fall back to '/' so the module evaluates and runs cleanly in the browser (where `url.fileURLToPath` is unavailable).
let __dirnameCache;
const getDirname = () => {
  if (__dirnameCache === undefined) {
    try {
      __dirnameCache = path.dirname(fileURLToPath(import.meta.url));
    } catch {
      __dirnameCache = '/';
    }
  }
  return __dirnameCache;
};

/**
 * Resolves a path relative to the wvFolder.
 * @param {string} wvFolder
 * @param {...string} segments
 * @returns {string}
 */
const wvResolve = (wvFolder, ...segments) => path.resolve(getDirname(), `../${wvFolder}`, ...segments);

/**
 * @typedef {{
 *   excludeOptimizedWorkers: 'y' | 'n',
 *   isNPMPackage: boolean,
 *   office: 'y' | 'n',
 *   pdfnetProd: 'y' | 'n',
 *   salesforceSupport: 'y' | 'n',
 *   useContentEdit: 'y' | 'n',
 *   keepFullAndLean: 'y' | 'n',
 *   useFullAPI: 'y' | 'n',
 *   useOfficeEditor: 'y' | 'n',
 *   useSpreadsheetEditor: 'y' | 'n',
 *   useSourceMap: 'y' | 'n',
 *   useWebComponent: 'y' | 'n',
 *   webViewerServer: 'y' | 'n',
 *   useLegacyOffice: 'y' | 'n',
 *   xod: 'y' | 'n',
 * }} Options
 */

/**
 * @typedef {{
 *   wvFolder: string,
 *   resourceDir: string,
 *   resourcesForZip: Array<string[]>,
 *   filesToDelete: string[],
 *   filesToDeleteAfterMoving: Array<{path: string, pattern: RegExp}>,
 *   filesToDeleteAsWildcard: Array<{path: string, pattern: RegExp}>,
 *   filesToRemoveSync: string[],
 *   filesToRename: Array<[string, string]>,
 *   filesToRelocate: Array<[string, string, boolean?]>,
 * }} FileManagementContext
 */


/**
 * Relocates package.json and webviewer.min.js for non-NPM packages.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleNPMPackageFiles = (options, ctx) => {
  if (options.isNPMPackage) {
    return;
  }
  const { wvFolder, resourceDir } = ctx;
  ctx.filesToRelocate.push(
    ['package.json', `${resourceDir}/${wvFolder}/`, true],
    ['webviewer.min.js', `${resourceDir}/${wvFolder}/`, true]
  );
};

/**
 * Deletes optimized worker folders when excluded.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleOptimizedWorkers = (options, ctx) => {
  if (options.excludeOptimizedWorkers !== 'y') {
    return;
  }
  ctx.filesToDelete.push(
    'core/pdf/lean/optimized',
    'core/pdf/full/optimized'
  );
};

/**
 * Handles legacyOffice files: delete or relocate based on useLegacyOffice option.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleLegacyOfficeFiles = (options, ctx) => {
  const { resourceDir } = ctx;

  if (options.useLegacyOffice === 'n') {
    ctx.filesToDelete.push('core/legacyOffice');
  } else {
    ctx.filesToRelocate.push(
      ['core/legacyOffice/LegacyOfficeWorker.js', `${resourceDir}/legacyOffice`],
      ['core/legacyOffice/WebB2XOfficeWorkerWasm.br.wasm', `${resourceDir}/legacyOffice`],
      ['core/legacyOffice/WebB2XOfficeWorkerWasm.br.js.mem', `${resourceDir}/legacyOffice`],
      ['core/legacyOffice/WebB2XOfficeWorker.br.js.mem', `${resourceDir}/legacyOffice_asm`],
      ['core/legacyOffice/WebB2XOfficeWorker.br.mem', `${resourceDir}/legacyOffice_resource`]
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/legacyOffice`],
      [`${resourceDir}/legacyOffice_asm`],
      [`${resourceDir}/legacyOffice_resource`]
    );
  }
};


/**
 * Handles office, legacyOffice, and PDF folders based on server/XOD/office options.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleOfficeAndServerFiles = (options, ctx) => {
  const { resourceDir } = ctx;

  if (options.webViewerServer === 'y') {
    ctx.filesToDelete.push(
      'core/office',
      'core/legacyOffice',
    );
  } else if (options.xod !== 'n') {
    ctx.filesToDelete.push(
      'core/office',
      'core/legacyOffice',
      'core/pdf'
    );
  } else if (options.office === 'n') {
    ctx.filesToDelete.push(
      'core/office',
      'core/legacyOffice'
    );
  } else if (options.office === 'y') {
    ctx.filesToRelocate.push(
      ['core/office/OfficeWorker.js', `${resourceDir}/office`],
      ['core/office/WebOfficeWorkerWasm.br.wasm', `${resourceDir}/office`],
      ['core/office/WebOfficeWorkerWasm.br.js.mem', `${resourceDir}/office`],
      ['core/office/WebOfficeWorker.br.js.mem', `${resourceDir}/office_asm`],
      ['core/office/WebOfficeWorker.br.mem', `${resourceDir}/office_resource`]
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/office`],
      [`${resourceDir}/office_asm`],
      [`${resourceDir}/office_resource`]
    );
    handleLegacyOfficeFiles(options, ctx);
  }
};

/**
 * Handles PDFNet.prod.js replacement when pdfnetProd option is enabled.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handlePDFNetProd = (options, ctx) => {
  if (options.pdfnetProd === 'y') {
    // replace PDFNet.js with PDFNet.prod.js
    ctx.filesToRemoveSync.push('core/pdf/PDFNet.js');
    ctx.filesToRename.push([
      'core/pdf/PDFNet.prod.js',
      'core/pdf/PDFNet.js',
    ]);
  } else {
    ctx.filesToRemoveSync.push('core/pdf/PDFNet.prod.js');
  }
};

/**
 * Handles full vs lean PDF API files based on keepFullAndLean and useFullAPI options.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handlePDFAPIFiles = (options, ctx) => {
  const { resourceDir } = ctx;

  if (options.keepFullAndLean === 'y') {
    ctx.filesToRelocate.push(
      ['core/pdf/full/optimized/PDFNetCWasm.br.js.mem', `${resourceDir}/pdf_full/full/optimized`],
      ['core/pdf/full/PDFNetC.br.mem', `${resourceDir}/resource/full`],
      ['core/pdf/lean/optimized/PDFNetCWasm.br.js.mem', `${resourceDir}/pdf_lean/lean/optimized`],
      ['core/pdf/lean/PDFNetC.br.mem', `${resourceDir}/resource/lean`]
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/pdf_lean`],
      [`${resourceDir}/pdf_full`]
    );
    handlePDFNetProd(options, ctx);
    return;
  }

  if (options.useFullAPI === 'n') {
    ctx.filesToDelete.push('core/pdf/full');
    ctx.filesToRelocate.push(
      ['core/pdf/lean/optimized/PDFNetCWasm.br.js.mem', `${resourceDir}/pdf_lean/lean/optimized`],
      ['core/pdf/lean/PDFNetC.br.mem', `${resourceDir}/resource/lean`]
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/pdf_lean`]
    );
  } else {
    ctx.filesToDelete.push('core/pdf/lean');
    handlePDFNetProd(options, ctx);
    ctx.filesToRelocate.push(
      ['core/pdf/full/optimized/PDFNetCWasm.br.js.mem', `${resourceDir}/pdf_full/full/optimized`],
      ['core/pdf/full/PDFNetC.br.mem', `${resourceDir}/resource/full`]
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/pdf_full`]
    );
  }
};

/**
 * Handles contentEdit files: delete or relocate based on useContentEdit option.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleContentEditFiles = (options, ctx) => {
  const { resourceDir } = ctx;

  if (options.useContentEdit === 'n') {
    ctx.filesToDelete.push('core/contentEdit');
  } else {
    ctx.filesToRelocate.push(
      ['core/contentEdit/InfixServerWasm.br.js.mem', `${resourceDir}/content_edit`],
      ['core/contentEdit/InfixServerWasm.br.wasm', `${resourceDir}/content_edit`],
      ['core/contentEdit/InfixServerWasm.gz.js.mem', `${resourceDir}/content_edit`],
      ['core/contentEdit/InfixServerModule.js', `${resourceDir}/content_edit`],
      ['core/contentEdit/InfixServerWasm.br.mem', `${resourceDir}/content_edit_resource`]
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/content_edit`],
      [`${resourceDir}/content_edit_resource`]
    );
  }
};

/**
 * Handles officeEditor files: delete or relocate based on useOfficeEditor option.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleOfficeEditorFiles = (options, ctx) => {
  const { resourceDir } = ctx;

  if (options.useOfficeEditor === 'n') {
    ctx.filesToDelete.push('core/officeEditor');
  } else {
    ctx.filesToRelocate.push(
      ['core/officeEditor/OfficeEditorWorkerWasm.br.js.mem', `${resourceDir}/office_edit`],
      ['core/officeEditor/OfficeEditorWorkerWasm.br.wasm', `${resourceDir}/office_edit`],
      ['core/officeEditor/OfficeEditorModule.js', `${resourceDir}/office_edit`]
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/office_edit`]
    );
  }
};

/**
 * Handles spreadsheetEditor files: delete or relocate based on useSpreadsheetEditor option.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleSpreadsheetEditorFiles = (options, ctx) => {
  const { resourceDir } = ctx;

  if (options.useSpreadsheetEditor === 'n') {
    ctx.filesToDelete.push('core/spreadsheetEditor');
  } else {
    ctx.filesToRelocate.push(
      ['core/spreadsheetEditor/SpreadsheetEditorEngine.js', `${resourceDir}/spreadsheetEditor`],
      ['core/spreadsheetEditor/SpreadsheetEditorEngine.wasm', `${resourceDir}/spreadsheetEditor`],
      ['core/spreadsheetEditor/SpreadsheetEditorModule.js', `${resourceDir}/spreadsheetEditor`],
      ['core/spreadsheetEditor/SpreadsheetEditor.css', `${resourceDir}/spreadsheetEditor`],
      ['core/spreadsheetEditor/styles.css', `${resourceDir}/spreadsheetEditor`],
      ['core/spreadsheetEditor/jquery-3.7.1.min.js', `${resourceDir}/spreadsheetEditor`],
      ['core/spreadsheetEditor/jquery.mousewheel-3.1.13.min.js', `${resourceDir}/spreadsheetEditor`],
      ['core/spreadsheetEditor/SpreadsheetEditor.js', `${resourceDir}/spreadsheetEditor`],
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/spreadsheetEditor`],
    );
  }
};

/**
 * Handles Salesforce-specific file management (fonts, external resources).
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleSalesforceFiles = (options, ctx) => {
  const { wvFolder, resourceDir } = ctx;

  if (options.salesforceSupport === 'y') {
    ctx.filesToRelocate.push(
      ['ui/assets/fonts', `${resourceDir}/font_assets`]
    );
    ctx.resourcesForZip.push(
      [`${resourceDir}/font_assets`]
    );
    ctx.filesToDeleteAfterMoving.push(
      { path: path.resolve(getDirname(), `${resourceDir}/lib/ui/assets/fonts/webfonts`), pattern: /.ttf/ },
      { path: path.resolve(getDirname(), `${resourceDir}/lib/ui/assets/fonts`), pattern: /.woff/ } 
    );
    ctx.resourcesForZip.push([`${resourceDir}/${wvFolder}`]);
  } else {
    ctx.filesToRelocate.push(
      ['core/external', `${resourceDir}/${wvFolder}/core/`, true]
    );
  }
};

/**
 * Handles source map file deletion based on useSourceMap and salesforceSupport options.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleSourceMapFiles = (options, ctx) => {
  if (options.useSourceMap !== 'n' && options.salesforceSupport !== 'y') {
    return;
  }
  ctx.filesToDelete.push(
    'ui/webviewer-ui.min.js.map',
    'ui/style.css.map'
  );
  ctx.filesToDeleteAsWildcard.push(
    { path: 'ui/chunks/', pattern: /\.(js|css)\.map$/ }
  );
};

/**
 * Handles web component HTML file selection based on useWebComponent option.
 * @param {Options} options
 * @param {FileManagementContext} ctx
 */
const handleWebComponentFiles = (options, ctx) => {
  if (options.useWebComponent === 'n') {
    ctx.filesToDelete.push('ui/index-wc.html');
  } else {
    ctx.filesToDelete.push('ui/index.html');
  }
};

/**
 * Optimizes file management based on the provided options.
 *
 * @param {Options} options - Configuration options for file management.
 * @param {string} resourceDir - Directory where resources will be stored.
 * @param {string} wvFolder - WebViewer folder name.
 * @returns {Object} An object containing arrays of files to delete, relocate, rename, etc.
 * @returns {Array} return.filesToDelete - List of files to delete.
 * @returns {Array} return.filesToDeleteAfterMoving - List of files to delete after moving.
 * @returns {Array} return.filesToDeleteAsWildcard - List of files to delete using wildcard patterns.
 * @returns {Array} return.filesToRelocate - List of files to relocate.
 * @returns {Array} return.filesToRemoveSync - List of files to remove synchronously.
 * @returns {Array} return.filesToRename - List of files to rename.
 * @returns {Array} return.resourcesForZip - List of resources to include in the zip.
 */
const optimizeFileManagement = (options, resourceDir, wvFolder) => {
  options = removeEmptyStringOptions(options);

  const ctx = {
    wvFolder,
    resourceDir,
    resourcesForZip: [],
    filesToDelete: [],
    filesToDeleteAfterMoving: [],
    filesToDeleteAsWildcard: [],
    filesToRemoveSync: [],
    filesToRename: [],
    filesToRelocate: [],
  };

  handleNPMPackageFiles(options, ctx);
  handleOptimizedWorkers(options, ctx);
  handleOfficeAndServerFiles(options, ctx);
  handlePDFAPIFiles(options, ctx);
  handleContentEditFiles(options, ctx);
  handleOfficeEditorFiles(options, ctx);
  handleSpreadsheetEditorFiles(options, ctx);
  handleSalesforceFiles(options, ctx);
  handleSourceMapFiles(options, ctx);
  handleWebComponentFiles(options, ctx);

  const resolve = (relativePath) => wvResolve(wvFolder, relativePath);

  return {
    filesToDelete: ctx.filesToDelete.map(resolve),
    filesToDeleteAfterMoving: ctx.filesToDeleteAfterMoving,
    filesToDeleteAsWildcard: ctx.filesToDeleteAsWildcard.map(({ path: p, pattern }) => ({
      path: resolve(p),
      pattern,
    })),
    filesToRelocate: ctx.filesToRelocate.map(([src, ...rest]) => [resolve(src), ...rest]),
    filesToRemoveSync: ctx.filesToRemoveSync.map(resolve),
    filesToRename: ctx.filesToRename.map(([oldPath, newPath]) => [resolve(oldPath), resolve(newPath)]),
    resourcesForZip: ctx.resourcesForZip,
  };
};

const removeEmptyStringOptions = (options) => {
  return Object.keys(options).reduce((acc, key) => {
    if (key === 'isNPMPackage') {
      acc[key] = options[key];
    } else {
      acc[key] = options[key] ? options[key] : 'n';
    }
    return acc;
  }, {});
};

export { optimizeFileManagement };
