// @pdftron/webviewer/webviewer.esm.js
/** Notice * This file contains works from many authors under various (but compatible) licenses. Please see core.txt for more information. **/
var e = {
	ASM: "asm",
	WASM: "ems",
	JS_WORKER: "jsworker",
	THREADED_WASM: "wasm-threads"
};
//#endregion
//#region src/webviewer/src/webviewerHelper.ts
function t(e, t) {
	if (typeof e != "string" || !t) return e;
	let n = t.substr(0, t.lastIndexOf("/"));
	return /^(\/|%2F|[a-z0-9-]+:)/i.test(e) ? e : `${n}/${e}`;
}
//#endregion
//#region src/webviewer/src/webviewer-wc.js
var n = 0, r = null, i = Promise.resolve(), a = 100;
function o(e, t) {
	return /^[a-z][a-z\d+.-]*:|^\/\//i.test(t) ? t : `${e?.endsWith("/") ? e.slice(0, -1) : e || ""}/${t?.startsWith("/") ? t.slice(1) : t}`;
}
function s(e) {
	try {
		let t = new URL(e, window.location.href);
		return `${t.origin}${t.pathname}`;
	} catch {
		return e.split(/[?#]/)[0];
	}
}
var c = () => {
	let e = localStorage.getItem("wv_instanceIds");
	if (e = e ? e.split(",") : [], e.length && e[n]) {
		let t = e[n];
		return n++, t;
	}
	let t = Math.random().toString(36).slice(-6);
	return e.findIndex((e) => e === t) === -1 ? (e.push(t), localStorage.setItem("wv_instanceIds", e), n++, t) : c();
}, l = class extends HTMLElement {
	connectedCallback() {
		if (p(), this._pendingDestroy && (clearTimeout(this._pendingDestroy), this._pendingDestroy = null), this.shadowRoot) {
			m(this.shadowRoot), window.dispatchEvent(new CustomEvent("apryse-webcomponent-reconnected", { detail: { host: this } }));
			return;
		}
		this.ensureShadowRoot(), this.getRuntimeState().isLoaded ? this.handleExistingRuntime() : this.handleFirstLoad();
	}
	ensureShadowRoot() {
		this.shadowRoot || m(this.attachShadow({ mode: "open" }));
	}
	getRuntimeState() {
		return {
			isLoaded: !!window.__apryseWebViewerUIRuntimeLoaded || !!window.webpackJsonpWebViewerUI,
			hasRemountEntrypoint: typeof window.__apryseWebViewerUIRemount == "function"
		};
	}
	handleFirstLoad() {
		this.injectScripts();
	}
	handleExistingRuntime() {
		this.injectScripts();
	}
	injectScripts() {
		let e = this.shadowRoot, t = this.getAttribute("path"), n = this.getAttribute("uiPath");
		r === null && (r = window.Core ? Promise.resolve() : new Promise((e, n) => {
			let i = document.createElement("script");
			i.type = "module", i.src = o(t, "core/webviewer-core.min.js"), i.onload = () => e(), i.onerror = () => {
				r = null, n(/* @__PURE__ */ Error(`Failed to load WebViewer Core script from ${i.src}`));
			}, window.document.head.appendChild(i);
		})), window.__apryseWebComponentMode = !0;
		let a = "ui/index-wc.html";
		n && (a = n), window.isApryseWebViewerWebComponent = !0, this.id = this.getAttribute("id") || c() || "default", i = i.then(() => u(this, e, t, a)).catch((e) => {
			console.error("[WebViewer WC] Failed to create instance:", e), this.dispatchEvent(new CustomEvent("webviewererror", { detail: { error: e } }));
		});
	}
	disconnectedCallback() {
		this._pendingDestroy = setTimeout(() => {
			this._pendingDestroy = null, this._destroyUI && (this._destroyUI(), this._destroyUI = null), this._removeMessageHandler && (this._removeMessageHandler(), this._removeMessageHandler = null), this.shadowRoot && h(this.shadowRoot);
		}, 0);
	}
};
async function u(e, t, n, i) {
	window.webViewerPath = n;
	let a = o(n, i), s = await fetch(a);
	if (!s.ok) throw Error(`Failed to fetch WebViewer template from ${a}: ${s.status} ${s.statusText}`);
	t.innerHTML = await s.text(), m(t), g(t, _(n, i)), f(t), await r, await d(t, n), window.createUIInstance(t), window.destroyUIInstance && (e._destroyUI = () => window.destroyUIInstance(t)), await new Promise((t) => {
		e.addEventListener("ready", t, { once: !0 });
	}), window.dispatchEvent(new Event("webviewerloaded"));
	let c = window.apryseTapPolyfill;
	c && typeof c.attachPolyfillTapEventsToWebComponent == "function" && c.attachPolyfillTapEventsToWebComponent(e);
}
function d(e, t) {
	let n = e.querySelectorAll("script"), r = [];
	return n.forEach((e) => {
		let n = document.createElement("script");
		for (let r of e.attributes) r.name === "src" ? n.setAttribute("src", o(t, r.value)) : n.setAttribute(r.name, r.value);
		n.text = e.innerHTML, n.async = !1, n.hasAttribute("src") && r.push(new Promise((e, t) => {
			n.onload = e, n.onerror = t;
		})), e.parentNode.replaceChild(n, e);
	}), Promise.all(r);
}
function f(e) {
	let t = document.getElementsByTagName("apryse-webviewer"), n = null;
	for (let r of t) if (r.shadowRoot && r.shadowRoot !== e) {
		n = r.shadowRoot.querySelector("head") || r.shadowRoot;
		break;
	}
	if (!n) return;
	let r = e.querySelector("head") || e, i = /* @__PURE__ */ new Set();
	r.querySelectorAll("link[rel=\"stylesheet\"]").forEach((e) => {
		let t = e.getAttribute("href");
		t && i.add(t);
	}), n.querySelectorAll("link[rel=\"stylesheet\"]").forEach((e) => {
		let t = e.getAttribute("href");
		t && !i.has(t) && r.appendChild(e.cloneNode(!0));
	});
}
function p() {
	if (window.__apryseInsertCss) return;
	let e = /* @__PURE__ */ new Set(), t = /* @__PURE__ */ new Map();
	window.__apryseRegisterShadowRoot = (n) => {
		if (!n) return;
		let r = n.querySelector("head") || n, i = /* @__PURE__ */ new Set();
		if (r.querySelectorAll("link[rel=\"stylesheet\"]").forEach((e) => {
			let t = e.getAttribute("href");
			t && i.add(t);
		}), e.size) {
			let t = e.values().next().value;
			(t.querySelector("head") || t).querySelectorAll("link[rel=\"stylesheet\"]").forEach((e) => {
				let t = e.getAttribute("href");
				if (!t || i.has(t)) return;
				let n = e.cloneNode(!0);
				r.appendChild(n), i.add(t);
			});
		}
		for (let e of t.values()) if (!i.has(e.href)) {
			let t = document.createElement("link");
			t.rel = "stylesheet", t.href = e.href;
			for (let [n, r] of Object.entries(e.attributes)) t.setAttribute(n, r);
			r.appendChild(t), i.add(e.href);
		}
		e.add(n);
	}, window.__apryseUnregisterShadowRoot = (t) => {
		t && e.delete(t);
	}, window.__apryseInsertCss = (n, r) => {
		let i = r || n.getAttribute("href") || n.href, o = {};
		for (let e of n.attributes) e.name !== "rel" && e.name !== "href" && (o[e.name] = e.value);
		if (i && (t.set(s(i), {
			href: i,
			attributes: o
		}), t.size > a && t.delete(t.keys().next().value)), !e.size) {
			document.getElementsByTagName("head")[0].appendChild(n);
			return;
		}
		let c = Array.from(e);
		(c[0].querySelector("head") || c[0]).appendChild(n);
		for (let e = 1; e < c.length; e++) {
			let t = c[e], i = t.querySelector("head") || t, a = n.cloneNode(!0);
			r && a.setAttribute("href", r), i.appendChild(a);
		}
	};
}
function m(e) {
	window.__apryseRegisterShadowRoot && window.__apryseRegisterShadowRoot(e);
}
function h(e) {
	window.__apryseUnregisterShadowRoot && window.__apryseUnregisterShadowRoot(e);
}
function g(e, t) {
	if (!t) return;
	let n = document.createElement("link");
	n.setAttribute("rel", "stylesheet"), n.setAttribute("href", `${t}`), (typeof e.querySelector == "function" && e.querySelector("head") || e).appendChild(n);
}
function _(e, t) {
	let n = e?.endsWith("/") ? e.slice(0, -1) : e || "", r = t?.includes("/") ? t.slice(0, t.lastIndexOf("/")) : "";
	return `${n}${r ? `/${r}` : ""}/style.css`;
}
try {
	customElements.define("apryse-webviewer", l);
} catch (e) {
	console.warn(`This environment does not allow for usage of the WebComponent version of WebViewer: ${e}`);
}
var v = function(e, t) {
	return Array.isArray(t) ? Promise.all(t.map((t) => C(e, t))) : C(e, t);
}, y = [
	"autoCreate",
	"autoFocusNoteOnAnnotationSelection",
	"enableOptimizedWorkers",
	"enableAnnotations",
	"preloadPDFWorker",
	"useDownloader",
	"showInvalidBookmarks"
], b = new Set(["disableLogs"]);
function x(e, t) {
	y.forEach((e) => {
		t?.[e] === void 0 && (t[e] = !0);
	});
	for (let n in t) {
		let r = t?.[n];
		if (r != null) {
			if (b.has(n) && r === !1) e.removeAttribute(n);
			else if (n === "encryption") e.setAttribute("auto_load", !1);
			else if (n !== "licenseKey" && n !== "l") {
				let t = Array.isArray(r) ? JSON.stringify(r) : r;
				e.setAttribute(n, t);
			}
		}
	}
}
function S(e, t) {
	return function() {
		return new Promise((n) => {
			F(e), t._removeMessageHandler && (t._removeMessageHandler(), t._removeMessageHandler = null), console.clear();
			let r = t.instance?.Core?.documentViewer?.closeDocument?.() || Promise.resolve();
			Promise.resolve(r).finally(() => {
				t._destroyUI && (t._destroyUI(), t._destroyUI = null), n();
			});
		});
	};
}
function C(e, t) {
	return new Promise((n) => {
		let r = document.createElement("apryse-webviewer");
		r.setAttribute("id", `wc-${t.id}`);
		let i = { ...e };
		i.l = i.l || i.licenseKey, i.azureWorkaround = i.azureWorkaround || i.enableAzureWorkaround, i.annotationAdmin = i.annotationAdmin || i.isAdminUser, i.enableReadOnlyMode = i.enableReadOnlyMode || i.isReadOnly, i.showLocalFilePicker = i.showLocalFilePicker || i.enableFilePicker, i.useSharedWorker = i.workerTransportPromise ? "true" : "false", i.path = i.path || "/lib", i.officeEditorOptions = JSON.stringify(i.officeEditorOptions || {}), i.spreadsheetEditorOptions = JSON.stringify(i.spreadsheetEditorOptions || {}), i.workerTransportPromise && (window.apryseWorkerTransportPromise = i.workerTransportPromise), i.l && (window.apryse_l = btoa(i.l)), x(r, i), r.setAttribute("webviewerjsversion", "12.0.0");
		let a = (e) => {
			e.data === "requestl" && e.source && e.source.postMessage({
				type: "responsel",
				value: i.l || i.licenseKey
			}, "*");
		};
		t.appendChild(r), r.addEventListener("ready", function() {
			if (i.config) {
				let e = document.createElement("script");
				e.src = i.config, document.body.appendChild(e);
			}
			if (i.encryption) {
				let e = i.initialDoc, t = {
					decrypt: r.instance.Core.Encryption.decrypt,
					decryptOptions: i.encryption,
					documentId: i.documentId,
					extension: "xod"
				};
				r.instance.UI.loadDocument(e, t);
			}
			if (i.additionalTranslations) {
				let e = i.additionalTranslations;
				r.instance.UI.setTranslations(e.language, e.translations);
			}
			P(t, r.instance), r.instance.UI.dispose = S(t, r), typeof r.loadInitialDocument == "function" && r.loadInitialDocument(), n(r.instance);
		}), window.addEventListener("message", a, !1), r._removeMessageHandler = () => window.removeEventListener("message", a, !1);
	});
}
window.WebViewerWebComponent = v;
//#endregion
//#region src/webviewer/src/webviewer.js
var w = {}, T = 0;
typeof console > "u" && (window.console = {
	log: function() {},
	warn: function() {},
	error: function() {}
});
var E = {
	extend: function() {
		for (let e = 1; e < arguments.length; e++) {
			let t = Object.keys(arguments[e]);
			for (let n = 0; n < t.length; n++) arguments[0][t[n]] = arguments[e][t[n]];
		}
		return arguments[0];
	},
	getValues: function(e) {
		let t = [];
		return e.forEach((e) => {
			t.push(e);
		}), t;
	},
	createEvent: function(e, t) {
		let n;
		try {
			n = new CustomEvent(e, {
				detail: t,
				bubbles: !0,
				cancelable: !0
			});
		} catch {
			n = document.createEvent("Event"), n.initEvent(e, !0, !0), n.detail = t;
		}
		return n;
	}
};
window.PDFNet && !w.skipPDFNetWebViewerWarning && console.warn("PDFNet.js and WebViewer.js have been included in the same context. See https://community.apryse.com/t/what-is-the-best-location-to-call-pdfnet-apis-when-used-in-conjunction-with-webviewer/5210 for an explanation of why this could be an error in your application.");
var D = /* @__PURE__ */ new Map(), O = /* @__PURE__ */ new Map();
w.WebViewer = function(e, t) {
	if (O.get(t)) throw Error("Two instances of WebViewer were created on the same HTML element. Please create a new element for each instance of WebViewer");
	{
		O.set(t, !0);
		let e = () => {
			let n = D.get(t);
			n.instance = this.getCompleteInstance(), t.removeEventListener("ready", e);
		};
		t.addEventListener("ready", e);
	}
	this._validateOptions(e);
	let n = e.webviewerServerURL || e.pdftronServer;
	e.fullAPI && n && (e.forceClientSideInit || console.warn("The fullAPI and webviewerServerURL options have both been set so the forceClientSideInit option is now enabled. This means that WebViewer will switch to client side rendering and processing to allow use of the full API."), e.forceClientSideInit = !0), this.options = E.extend({}, w.WebViewer.Options, e);
	let r = this.options.path.length - 1;
	r > 0 && this.options.path[r] !== "/" && (this.options.path += "/"), this.options.uiPath = this.options.path + this.options.uiPath, t || console.error("ViewerElement is not defined. This may be caused by calling the WebViewer constructor before the DOM is loaded, or an invalid selector. Please see https://docs.apryse.com/documentation/web/get-started/ for an example."), this.element = t, this.element.style.overflow = "hidden";
	let i = this;
	this.messageHandler = function(t) {
		if (t.data === "requestl" && t.source && t.source.postMessage({
			type: "responsel",
			value: e.l || e.licenseKey
		}, "*"), t.data.type === "requestConfig" && t.data.id === i.id && t.source) {
			let n = e.config ? i._correctRelativePath(e.config) : "";
			t.source.postMessage({
				type: "responseConfig",
				value: n
			}, "*");
		}
	}, window.addEventListener("message", this.messageHandler, !1), this.options.autoCreate && this.create();
};
var k = {
	licenseKey: void 0,
	enableAzureWorkaround: !1,
	isAdminUser: !1,
	isReadOnly: !1
};
w.WebViewer.prototype = {
	version: "12.0.0",
	create: function() {
		if (this.options.initialDoc) {
			let e = this._correctRelativePath(this.options.initialDoc);
			e = encodeURIComponent(JSON.stringify(e)), this.options.initialDoc = e;
		}
		this._create();
	},
	_create: function() {
		this.id = ++T, this._trigger === void 0 && (this._trigger = function(e, t) {
			let n = E.createEvent(e, t);
			this.element.dispatchEvent(n);
		});
		let e = this.options.type.replace(" ", "").split(",");
		e.length < 1 && (e[0] = "html5"), this._createViewer(e);
	},
	_validateOptions: function(e) {
		let t = Object.keys(e);
		for (let e = 0; e < t.length; e++) {
			let n = t[e];
			n in E.extend({}, k, w.WebViewer.Options) || console.warn(`${n} is not a valid option name. See https://docs.apryse.com/api/web/global.html#WebViewerOptions for all available options.`);
		}
		let n = e.webviewerServerURL || e.pdftronServer;
		e.enableRedaction && !(e.fullAPI || n) && console.warn("FullAPI or WebViewer Server is needed to apply redactions");
	},
	_notSupportedMobile: function() {
		let e = document.createElement("div");
		e.id = "webviewer-browser-unsupported", e.textContent = "PDF document viewing is not supported by this browser.", this.element.appendChild(e);
	},
	_createViewer: function(e) {
		let t = this;
		t.selectedType = null;
		let n;
		if (this.isMobileDevice()) {
			if (this.options.documentType && this.options.documentType !== "xod" && !this._testWebAssembly()) {
				this._notSupportedMobile();
				return;
			}
			if (e = ["html5Mobile"], t.selectedType = "html5Mobile", this.options.mobileRedirect) {
				n = this.options.uiPath + this._getHTML5OptionsURL(), window.location = n;
				return;
			}
		}
		let r = !1, i = !1;
		for (let a = 0; a < e.length; a++) {
			if (e[a].toLowerCase() === "html5mobile") {
				if (this.options.documentType && this.options.documentType !== "xod" && !this._testWebAssembly()) continue;
				if (r = !0, t._testHTML5()) {
					if (this.options.mobileRedirect) {
						t.selectedType = "html5Mobile", n = this.options.uiPath + this._getHTML5OptionsURL(), window.location = n;
						return;
					}
					if (this.options.xdomainProxyUrl || t.isSameOrigin(decodeURIComponent(t.options.initialDoc)) || t._testCORS()) {
						t.selectedType = "html5Mobile";
						break;
					} else i = !0;
				}
			}
			if (e[a].toLowerCase() === "html5" && (r = !0, t._testHTML5())) {
				let e = t.isSameOrigin(decodeURIComponent(t.options.initialDoc));
				if (this.options.xdomainProxyUrl || e || t._testCORS()) {
					t.selectedType = "html5";
					break;
				} else i = !0;
			}
		}
		if (t.selectedType === "html5") t._createHTML5();
		else if (t.selectedType === "html5Mobile") t._createHTML5Mobile();
		else {
			let e;
			if (i ? e = "This browser does not support cross origin requests. Please configure xdomain to support CORS." : r && (e = "Please use an HTML5 compatible browser."), e) {
				let n = document.createElement("div");
				n.id = "webviewer-browser-unsupported", n.textContent = e, t.element.appendChild(n);
			}
		}
	},
	_viewerLoaded: function(e) {
		this._trigger("ready");
		try {
			let t = e.contentWindow;
			if (t.postMessage({ type: "viewerLoaded" }, "*"), this.options.encryption !== void 0) {
				let e = decodeURIComponent(this.options.initialDoc);
				this.options.initialDoc && (e = JSON.parse(e));
				let n = {
					decrypt: t.Core.Encryption.decrypt,
					decryptOptions: this.options.encryption,
					documentId: this.options.documentId,
					extension: "xod"
				};
				this.getInstance().UI.loadDocument(e, n);
			}
		} catch {
			console.warn("Viewer is on a different domain, the promise from WebViewer function is rejected and API functions will not work because of cross domain permissions. See https://docs.apryse.com/documentation/web/guides/wv-inside/#loading-webviewer-from-another-domain for more information.");
		}
	},
	_getHTML5OptionsURL: function() {
		let e = this.options, t = e.webviewerServerURL || e.pdftronServer, n = "";
		e.initialDoc && (n += `#d=${e.initialDoc}`), e.backendType === void 0 && (e.backendType = e.pdfBackend);
		let r, i;
		if (e.documentType && e.documentType === "xod" && (r = e.documentType), e.preloadWorker && e.preloadWorker === "xod" && (r = e.preloadWorker), e.extension && (r = e.extension), r && (n += `&extension=${r}`), e.documentType && e.documentType !== "xod" && (i = e.documentType), e.preloadWorker && e.preloadWorker !== "xod" && (i = e.preloadWorker), i && (n += `&preloadWorker=${i}`), e.backendType && (n += `&pdf=${e.backendType}&office=${e.backendType}&legacyOffice=${e.backendType}`), e.filename && (n += `&filename=${e.filename}`), e.streaming !== void 0 && (n += `&streaming=${e.streaming}`), e.externalPath) {
			let t = this._correctRelativePath(e.externalPath);
			t = encodeURIComponent(t), n += `&p=${t}`;
		}
		if (e.encryption && (n += "&auto_load=false"), e.enableAnnotations && (n += "&a=1"), e.disabledElements) {
			let t = encodeURIComponent(e.disabledElements.join(","));
			n += `&disabledElements=${t}`;
		}
		if (e.serverUrl) {
			let t = this._correctRelativePath(e.serverUrl);
			t = encodeURIComponent(t), n += `&server_url=${t}`;
		}
		if (e.serverUrlHeaders && (n += `&serverUrlHeaders=${JSON.stringify(e.serverUrlHeaders)}`), e.documentId && (n += `&did=${e.documentId}`), e.css) {
			let t = this._correctRelativePath(e.css);
			t = encodeURIComponent(t), n += `&css=${t}`;
		}
		if (e.disableI18n && (n += "&disableI18n=1"), e.enableOfflineMode && (n += "&offline=1"), e.startOffline && (n += "&startOffline=1"), (e.enableReadOnlyMode || e.isReadOnly) && (n += "&readonly=1"), e.hideAnnotationPanel && (n += "&hideAnnotationPanel=1"), e.disableToolGroupReordering && (n += "&disableToolGroupReordering=1"), e.annotationUser !== void 0 && (n += `&user=${e.annotationUser}`), (e.annotationAdmin !== void 0 || e.isAdminUser !== void 0) && (n += `&admin=${e.annotationAdmin || e.isAdminUser ? 1 : 0}`), e.custom !== void 0 && (n += `&custom=${encodeURIComponent(e.custom)}`), (e.showLocalFilePicker !== void 0 || e.enableFilePicker !== void 0) && (n += `&filepicker=${e.showLocalFilePicker || e.enableFilePicker ? 1 : 0}`), e.fullAPI !== void 0 && (n += `&pdfnet=${+!!e.fullAPI}`), e.enableRedaction !== void 0 && (n += `&enableRedaction=${+!!e.enableRedaction}`), e.disableVirtualDisplayMode !== void 0 && (n += `&disableVirtualDisplayMode=${+!!e.disableVirtualDisplayMode}`), e.hideDetachedReplies !== void 0 && (n += `&hideDetachedReplies=${+!!e.hideDetachedReplies}`), e.enableMeasurement !== void 0 && (n += `&enableMeasurement=${+!!e.enableMeasurement}`), e.showToolbarControl !== void 0 && (n += `&toolbar=${e.showToolbarControl ? "true" : "false"}`), e.notesInLeftPanel !== void 0 && (n += `&notesInLeftPanel=${+!!e.notesInLeftPanel}`), e.autoExpandOutlines !== void 0 && (n += `&autoExpandOutlines=${+!!e.autoExpandOutlines}`), e.enableAnnotationNumbering !== void 0 && (n += `&enableAnnotationNumbering=${+!!e.enableAnnotationNumbering}`), e.enableOfficeEditing !== void 0 && (n += `&enableOfficeEditing=${+!!e.enableOfficeEditing}`), e.officeEditorOptions && (n += `&officeEditorOptions=${JSON.stringify(e.officeEditorOptions)}`), e.spreadsheetEditorOptions !== void 0 && (n += `&spreadsheetEditorOptions=${JSON.stringify(e.spreadsheetEditorOptions)}`), e.xdomainProxyUrl !== void 0) {
			let t;
			t = typeof e.xdomainProxyUrl == "string" ? { url: e.xdomainProxyUrl } : e.xdomainProxyUrl, n += `&xdomain_urls=${encodeURIComponent(JSON.stringify(t))}`;
		}
		return (e.azureWorkaround || e.enableAzureWorkaround) && (n += "&azureWorkaround=1"), e.enableOptimizedWorkers || (n += "&enableOptimizedWorkers=0"), e.useDownloader || (n += "&useDownloader=0"), e.disableWebsockets && (n += "&disableWebsockets=1"), e.defaultLanguage && (n += `&defaultLanguage=${e.defaultLanguage}`), e.disableStreaming && (n += "&disableStreaming=1"), e.forceClientSideInit && (n += "&forceClientSideInit=1"), e.loadAsPDF && (n += "&loadAsPDF=1"), e.workerTransportPromise !== void 0 && (n += `&useSharedWorker=${e.workerTransportPromise ? "true" : "false"}`), t !== void 0 && t && (n += `&webviewerServerURL=${encodeURIComponent(t)}`), e.fallbackToClientSide && (n += "&fallbackToClientSide=1"), e.singleServerMode !== void 0 && (n += `&singleServerMode=${e.singleServerMode ? "true" : "false"}`), e.webviewerServerRangeRequests !== void 0 && (n += `&wvsRange=${e.webviewerServerRangeRequests ? "true" : "false"}`), e.accessibleMode !== void 0 && (n += `&accessibleMode=${+!!e.accessibleMode}`), e.disableLogs && (n += "&disableLogs=1"), e.enableViewStateAnnotations && (n += "&enableViewStateAnnotations=1"), e.disableFlattenedAnnotations && (n += "&disableFlattenedAnnotations=1"), e.selectAnnotationOnCreation !== void 0 && (n += `&selectAnnotationOnCreation=${+!!e.selectAnnotationOnCreation}`), e.autoFocusNoteOnAnnotationSelection !== void 0 && (n += `&autoFocusNoteOnAnnotationSelection=${+!!e.autoFocusNoteOnAnnotationSelection}`), e.disableIndexedDB && (n += "&disableIndexedDB=1"), e.disableMultiViewerComparison && (n += "&disableMultiViewerComparison=1"), e.showInvalidBookmarks && (n += "&showInvalidBookmarks=1"), e.disableObjectURLBlobs && (n += `&disableObjectURLBlobs=${e.disableObjectURLBlobs}`), n += `&id=${this.id}`, n += `&basePath=${encodeURIComponent(window.location.pathname)}`, n.length > 0 && n[0] === "&" && (n = `#${n.slice(1)}`), n += `&webViewerJSVersion=${this.version}`, e.uiConfig && (n += `&uiConfig=${e.uiConfig}`), e.initialMode && (n += `&initialMode=${e.initialMode}`), e.chunkSize && (n += `&chunkSize=${e.chunkSize}`), e.enableCompositionInput !== "undefined" && (n += `&enableCompositionInput=${+!!e.enableCompositionInput}`), n;
	},
	setInstanceData: function(e) {
		D.set(this.element, {
			iframe: e,
			l: this.options.l || this.options.licenseKey,
			workerTransportPromise: this.options.workerTransportPromise
		});
	},
	_createHTML5: function() {
		let e = this.options.uiPath + this._getHTML5OptionsURL(), t = document.createElement("iframe");
		this.setInstanceData(t), t.id = `webviewer-${this.id}`, t.src = e, t.title = "webviewer", t.frameBorder = 0, t.width = "100%", t.height = "100%", t.setAttribute("allowfullscreen", !0), t.setAttribute("webkitallowfullscreen", !0), t.setAttribute("mozallowfullscreen", !0), this.iframe = t, this.options.backgroundColor && t.setAttribute("data-bgcolor", this.options.backgroundColor), this.options.assetPath && t.setAttribute("data-assetpath", encodeURIComponent(this.options.assetPath));
		let n = this.options.cspNonce;
		n && (t.dataset.cspNonce = n), this.element.appendChild(t);
		let r = (e) => {
			if (e.data.type === "viewerLoaded" && e.data.id === this.id) try {
				this.instance = this.iframe.contentWindow.instance;
			} catch {} finally {
				window.removeEventListener("message", r), this._viewerLoaded(this.iframe);
			}
		};
		window.addEventListener("message", r);
	},
	_createHTML5Mobile: function() {
		let e = this.options.uiPath + this._getHTML5OptionsURL(), t = document.createElement("iframe");
		this.setInstanceData(t), t.id = `webviewer-${this.id}`, t.src = e, t.frameborder = 0, this.options.assetPath && t.setAttribute("data-assetpath", encodeURIComponent(this.options.assetPath));
		let n = this.options.cspNonce;
		n && (t.dataset.cspNonce = n), t.style.width = "100%", t.style.height = "100%", this.iframe = t, this.element.appendChild(t);
		let r = (e) => {
			if (e.data.type === "viewerLoaded" && e.data.id === this.id) try {
				this.instance = this.iframe.contentWindow.instance;
			} catch {} finally {
				window.removeEventListener("message", r), this._viewerLoaded(this.iframe);
			}
		};
		window.addEventListener("message", r);
	},
	dispose: function() {
		return new Promise((e) => {
			D.delete(this.element), O.delete(this.element), window.removeEventListener("message", this.messageHandler), this.iframe = null, console.clear(), this.instance.Core.documentViewer.unmount().then(() => {
				e();
			});
		});
	},
	getInstance: function() {
		return this.instance;
	},
	setCompleteInstance: function(e) {
		this.completeInstance = e;
	},
	getCompleteInstance: function() {
		return this.completeInstance;
	},
	_correctRelativePath: function(e) {
		let n = window.location.pathname;
		return Array.isArray(e) ? e.map((e) => t(e, n)) : t(e, n);
	},
	_testHTML5: function() {
		try {
			let e = document.createElement("canvas");
			return e && e.getContext("2d");
		} catch {
			return !1;
		}
	},
	_testWebAssembly: function() {
		return !!(window.WebAssembly && window.WebAssembly.validate);
	},
	_testCORS: function() {
		return "XMLHttpRequest" in window && "withCredentials" in new XMLHttpRequest();
	},
	isIE: function() {
		let e = navigator.userAgent.toLowerCase(), t = /(msie) ([\w.]+)/.exec(e) || /(trident)(?:.*? rv:([\w.]+)|)/.exec(e);
		return t && parseInt(t[2], 10);
	},
	isMobileDevice: function() {
		return !this.isIE() && (this.scrollbarWidth() === 0 && navigator.userAgent.match(/Edge/i) || navigator.userAgent.match(/Android/i) || navigator.userAgent.match(/webOS/i) || navigator.userAgent.match(/iPhone/i) || navigator.userAgent.match(/iPod/i) || navigator.userAgent.match(/iPad/i) || navigator.userAgent.match(/Touch/i) || navigator.userAgent.match(/IEMobile/i) || navigator.userAgent.match(/Silk/i));
	},
	scrollbarWidth: function() {
		this._scrollbarSheet || (this._scrollbarSheet = new CSSStyleSheet(), this._scrollbarSheet.replaceSync(".apryse-scrollbar-measure { width: 100px; height: 100px; overflow: scroll; position: absolute; top: -9999px; }")), document.adoptedStyleSheets.includes(this._scrollbarSheet) || (document.adoptedStyleSheets = [...document.adoptedStyleSheets, this._scrollbarSheet]);
		let e = document.createElement("div");
		e.classList.add("apryse-scrollbar-measure"), document.body.appendChild(e);
		let t = e.offsetWidth - e.clientWidth;
		return document.body.removeChild(e), t;
	},
	isSameOrigin: function(e) {
		let t = window.location, n = document.createElement("a");
		n.href = e, n.host === "" && (n.href = n.href);
		let r = window.location.port, i = n.port;
		return n.protocol === "http:" ? (i = i || "80", r = r || "80") : n.protocol === "https:" && (i = i || "443", r = r || "443"), n.hostname === t.hostname && n.protocol === t.protocol && i === r;
	}
}, w.WebViewer.Options = {
	initialDoc: void 0,
	annotationAdmin: void 0,
	isAdminUser: void 0,
	annotationUser: void 0,
	assetPath: void 0,
	autoCreate: !0,
	autoFocusNoteOnAnnotationSelection: !0,
	azureWorkaround: !1,
	additionalTranslations: void 0,
	enableAzureWorkaround: !1,
	enableOptimizedWorkers: !0,
	backgroundColor: void 0,
	backendType: void 0,
	css: void 0,
	config: void 0,
	custom: void 0,
	documentId: void 0,
	documentType: void 0,
	preloadWorker: void 0,
	extension: void 0,
	enableAnnotations: !0,
	filename: void 0,
	disableI18n: !1,
	disabledElements: void 0,
	disableWebsockets: !1,
	enableOfflineMode: !1,
	enableReadOnlyMode: !1,
	isReadOnly: !1,
	enableRedaction: !1,
	disableVirtualDisplayMode: !1,
	enableMeasurement: !1,
	encryption: void 0,
	externalPath: void 0,
	hideAnnotationPanel: !1,
	disableToolGroupReordering: !1,
	uiPath: "./ui/index.html",
	l: void 0,
	licenseKey: void 0,
	mobileRedirect: !1,
	path: "",
	pdfBackend: void 0,
	webviewerServerURL: void 0,
	webviewerServerRangeRequests: !0,
	fallbackToClientSide: !1,
	singleServerMode: !1,
	fullAPI: !1,
	preloadPDFWorker: !0,
	serverUrl: void 0,
	serverUrlHeaders: void 0,
	showLocalFilePicker: !1,
	enableFilePicker: !1,
	showToolbarControl: void 0,
	startOffline: !1,
	streaming: void 0,
	type: "html5",
	useDownloader: !0,
	workerTransportPromise: void 0,
	xdomainProxyUrl: void 0,
	ui: void 0,
	uiConfig: void 0,
	forceClientSideInit: !1,
	loadAsPDF: !1,
	accessibleMode: void 0,
	disableLogs: !1,
	enableViewStateAnnotations: !1,
	selectAnnotationOnCreation: !1,
	notesInLeftPanel: !1,
	autoExpandOutlines: !1,
	enableAnnotationNumbering: !1,
	enableOfficeEditing: !1,
	officeEditorOptions: {},
	spreadsheetEditorOptions: {},
	documentXFDFRetriever: void 0,
	disableMultiViewerComparison: void 0,
	showInvalidBookmarks: !0,
	disableObjectURLBlobs: !1,
	initialMode: void 0,
	defaultLanguage: "en",
	chunkSize: void 0,
	enableCompositionInput: !1,
	cspNonce: void 0,
	disableFlattenedAnnotations: !1,
	disableStreaming: !1
};
var A = function(e, t) {
	return new Promise(function(n, r) {
		e.l = e.l || e.licenseKey, e.azureWorkaround = e.azureWorkaround || e.enableAzureWorkaround, e.annotationAdmin = e.annotationAdmin || e.isAdminUser, e.enableReadOnlyMode = e.enableReadOnlyMode || e.isReadOnly, e.showLocalFilePicker = e.showLocalFilePicker || e.enableFilePicker;
		let i = function() {
			t.removeEventListener("ready", i);
			try {
				let r = t.querySelector("iframe").contentWindow;
				if (r.Core.Tools === void 0) throw Error("Viewer isn't instantiated correctly. It could be caused by the 'path' option in the WebViewer constructor not being defined correctly. The 'path' option should be relative to the HTML file you're loading the script on and the lib folder needs to be publicly accessible.");
				let i = a.getInstance(), o = {
					iframeWindow: r,
					dispose: a.dispose.bind(a)
				}, s = E.extend({}, i, { [i.UI_NAMESPACE_KEY]: {
					...i[i.UI_NAMESPACE_KEY],
					...o
				} });
				a.setCompleteInstance(s);
				let c = Promise.resolve();
				if (e.documentXFDFRetriever && (c = s[i.CORE_NAMESPACE_KEY].documentViewer.setDocumentXFDFRetriever(e.documentXFDFRetriever)), e.additionalTranslations) {
					let t = e.additionalTranslations;
					s[i.UI_NAMESPACE_KEY].setTranslations(t.language, t.translations);
				}
				c.then(() => {
					n(s);
				});
			} catch {
				if ("config" in e && e.config !== "") return n();
				r("Viewer is on a different domain, the promise from WebViewer function is rejected and API functions will not work because of cross domain permissions. See https://docs.apryse.com/documentation/web/guides/wv-inside/#loading-webviewer-from-another-domain for more information.");
			}
		};
		t.addEventListener("ready", i);
		let a = new w.WebViewer(e, t);
	});
}, j = function(e) {
	let t = E.getValues(D);
	for (let n = 0; n < t.length; n++) {
		let r = t[n], { iframe: i } = r;
		if (i === e) return r;
	}
	return null;
}, M = v;
M.l = function(e) {
	let t = j(e);
	return t && t.l;
}, M.workerTransportPromise = function(e) {
	let t = j(e);
	return t && t.workerTransportPromise;
}, M.WorkerTypes = {
	ALL: "all",
	OFFICE: "office",
	LEGACY_OFFICE: "legacyOffice",
	PDF: "pdf",
	CONTENT_EDIT: "contentEdit",
	OFFICE_EDITOR: "officeEditor"
}, M.BackendTypes = e, M.Modes = {
	DEFAULT: "default",
	DOCX_EDITOR: "docxEditor",
	SPREADSHEET_EDITOR: "spreadsheetEditor"
}, M.InitialEditModes = {
	EDITING: "editing",
	REVIEWING: "reviewing",
	VIEW_ONLY: "viewOnly",
	PREVIEW: "preview"
};
var N = function(e) {
	let t = E.getValues(D);
	if (!t.length || !t.filter((e) => e.instance).length) return console.warn("WebViewer.getInstance() was called before any instances were available"), null;
	if (t.length > 1 && !e) throw Error("More than one instance of WebViewer was found, and no element was passed into getInstance(). Please specify which instance you want to get.");
	return e ? (D.get(e) || {}).instance : (t[0] || {}).instance;
}, P = function(e, t) {
	D.set(e, { instance: t });
}, F = function(e) {
	D.delete(e), O.delete(e);
};
M.getInstance = N, window.WebViewer = M, window.getInstance = N, M.WebComponent = M, M.Iframe = A;
//#endregion
export { F as clearInstanceForWebComponent, M as default, N as getInstance, P as setInstanceForWebComponent };
